---
title: "DNA Variant Analysis with R Bioconductor"
output: html_notebook
---

------------------------------------------------------------------------

#### *Objective 6: Export variants to a Browser Extensible Data (BED) file and view in UCSC browser*

------------------------------------------------------------------------

#### A. Select variants to include in the BED file

Browser extensible data (BED) files are simple text files that list variant chromosomes, positions, and sometimes dbSNP annotations (for SNPs), sometimes strand and other information. These are useful for direct loading into the UCSC Genome Browser, which allows you to create custom tracks showing your sample variants alongside other types of annotations.

Since we saw earlier that we have a few variants in non-TRP genes, we can use this as a small example. Let's subset these variants so we can visualize where these are in the UCSC browser, and their relation to the TRP genes. Recall from Objective 4, we created `other_ids`.

```{r}
table("gene loc:"=all_df$LOCATION, "ENTREZID:"=all_df$GENEID)

other_ids
```

------------------------------------------------------------------------

We have 2 variants in SHPK with ENTREZID 23729 so let's select those from `all_vars`. We can use regular R subsetting methods based to put these 2 into a `shpk_vars` object. Remember that `all_vars` is the full set of variants from our VCF file and the ENTREZIDs are stored in the column GENEID.

```{r}
table("ENTREZID:"=all_vars$GENEID)
```

SHPK corresponds to ENTREZID 23729 as we see from the `other_ids` table above, so we can use this identifier to create `shpk_vars`:

```{r warning=FALSE}
shpk_vars <- all_vars[which(all_vars$GENEID=="23729")]
shpk_vars
```

------------------------------------------------------------------------

#### B. Re-read VCF with selected fields and transcripts

Recall from Objective 3 that we can select specific columns (e.g. read depth, genotypes, and ploidy) and variants for certain genes with the `ScanVcfParam()` function, and then we can re-read the VCF with `readVcf()` function. Let's do this again for the SHPK variants:

```{r}
shpk_params <- ScanVcfParam(info="DP", geno=c("GT", "cPd"), which=shpk_vars)
shpk_params
shpk_vcf <- readVcf(vcf_file, "hg19", shpk_params)
shpk_vcf
rowRanges(shpk_vcf)
```

------------------------------------------------------------------------

##### Review specific variant information to export to BED

So we see that the first of these is a single nucleotide polymorphism (SNP) and the second one is a deletion of GAGA.

The ENSEMBL website has a nice summary of small variant types found in VCFs (<http://uswest.ensembl.org/info/genome/variation/index.html>):

------------------------------------------------------------------------

#### C. Set up a GRanges object for UCSC 

In order to view these in the UCSC browser, we need to set up a special GRanges object for this with just the chromosome (back to "chr"), ranges, strand, and variant types. You can do this with the `GRangesForUCSCGenome()` function, but as the help file for this indicates:

"`GRangesForUCSCGenome` obtains sequence information from the UCSC website, while `GRangesForBSGenome` looks for it in an installed `BSGenome` package. Using the latter is more efficient in the long-run, but requires downloading and installing a potentially large genome package, or creating one from scratch if it does not yet exist for the genome of interest."

So you need an internet connection to use the `GRangesForUCSCGenome` function. I will use the `GRangesForBSGenome` function instead since we have already used this.

```{r}
?GRangesForUCSCGenome

ranges(shpk_vars) ## same as shpk_vars@ranges (slot syntax)
strand(shpk_vars) ## same as shpk_vars@strand (slot syntax)

shpk_track <- GRangesForBSGenome(genome="hg19", 
    chrom = "chr17", 
    ranges = shpk_vars@ranges,
    strand = shpk_vars@strand)

shpk_track
```

Now that we have all the info that we want to include in our UCSC Genome track, we can export this.

------------------------------------------------------------------------

#### D. Export the data to UCSC Genome Browser

##### Save the BED file 

This will over-write the file that is on the course website, which is there as a backup. The `export()` function just writes it for you if you didn't already have it saved.

```{r}
export(shpk_track, paste0(proj_dir, "/Data/shpk.bed"))
```

------------------------------------------------------------------------

##### Import an existing BED file

If you are reading in an already-created BED file you can create `shpk_track` using the `import()` function.

```{r}
shpk_track <- import(paste0(proj_dir,"/Data/shpk.bed"))
shpk_track
```

------------------------------------------------------------------------

##### Open the BED file with the UCSC Genome Browser 

Finally, let's view the second variant (a deletion) in the UCSC browser, zooming out 3x at least so we can see a little of the surrounding area, and selecting just a few tracks.

This requires an internet connection! It might take a minute but it should open a new web browser once finished. Note that our dbSNP database selection I am using here is dbSNP 153. Our original VCF file was annotated with dbSNP 131, but the UCSC Browser no longer has this track. It is fine to a newer version of dbSNP here because we are just mapping on position. And you can select different versions from the browser.

```{r}
session <- browserSession("UCSC")
genome(session) <- "hg19"
track(session, "SHPK_variants") <- shpk_track

shpk_snp1 <- shpk_track[1]
shpk_snp1

## get all the various track names (lots of output... uncomment to view)
# trackNames(session)

view <- browserView(session, 
        shpk_snp1* -3, ## this line zooms it out 3x
        track=c("ct_SHPKvariants_7561", "ruler", "knownGene", "dbSNP 153"))
```

Note: you can also zoom in/out and select tracks from within the UCSC browser window until you get the view that you like, then you can update the code above.
